narwhal/consensus/src/utils.rs

```diff
@@ -55,15 +55,10 @@ fn linked(leader: &Certificate, prev_leader: &Certificate, dag: &Dag) -> bool {
 }
 
 /// Flatten the dag referenced by the input certificate. This is a classic depth-first search (pre-order):
-/// https://en.wikipedia.org/wiki/Tree_traversal#Pre-order
-pub fn order_dag(
-    gc_depth: Round,
-    leader: &Certificate,
-    state: &ConsensusState,
-) -> Vec<Certificate> {
+/// <https://en.wikipedia.org/wiki/Tree_traversal#Pre-order>
+pub fn order_dag(leader: &Certificate, state: &ConsensusState) -> Vec<Certificate> {
     debug!("Processing sub-dag of {:?}", leader);
     assert!(leader.round() > 0);
-    let gc_round = leader.round().saturating_sub(gc_depth);
```

Comment:
@mwtian this is the part that I am talking about the in the description of the PR.

I believe using the `leader round` as reference to calculate the `gc_round` when ordering the DAG is quite aggressive and will lead to always skipping rounds between 2 subdag commits, as we don't try to commit certificates up to the last known calculated  `gc_round`  (where we have legit history from that point onwards) and then progressively update the `gc_round`. 
Another way to think about it is on long periods of asynchrony where we might even have missing leaders in between two sub dag commits. So imagine the scenario where we committed last time on round `100` (so `gc_round == 50`), and next commit (successful leader election) happens on round `200` (no leaders are found in between). Now with the existing logic we will commit certificates up to round `200 - gc_depth = 200 - 50 + 1 = 151` , although we could commit up to round `100 + 1 = 101` without any issues.

Reply:
I don't think rounds will always be skipped. Skipping between commits will only happen if there are certificates above last GC round and the last committed round of its origin, but at or below the new GC round.

Personally I believe it is more robust to always traverse at most GC depth, instead of traversing more than GC depth when there are long periods of failed leaders. Similar to the example, if a leader is selected after e.g. 10min of no successful leader selection, do we want to commit a sub dag with ~ 50k certificates or more? That would seem quite disruptive to the system and not a tested scenario.

Reply:
My understanding was that the gc_depth means the dag has at most that many rounds, regardless of which ones committed. Did this change recently?